File: src/charencoding.cpp
#include "charencoding.h"

#include "libNewsboat-ffi/src/charencoding.rs.h"

namespace Newsboat {
namespace charencoding {

std::optional<std::string> charset_from_bom(std::vector<std::uint8_t> content)
{
	rust::String charset;
	const auto input = rust::Slice<const std::uint8_t>(content.data(), content.size());
	if (charencoding::bridged::charset_from_bom(input, charset)) {
		return std::string(charset);
	}
	return {};
}

std::optional<std::string> charset_from_xml_declaration(std::vector<std::uint8_t>
	content)
{
	rust::String charset;
	const auto input = rust::Slice<const std::uint8_t>(content.data(), content.size());
	if (charencoding::bridged::charset_from_xml_declaration(input, charset)) {
		return std::string(charset);
	}
	return {};
}

std::optional<std::string> charset_from_content_type_header(std::vector<std::uint8_t>
	header)
{
	rust::String charset;
	const auto input = rust::Slice<const std::uint8_t>(header.data(), header.size());
	if (charencoding::bridged::charset_from_content_type_header(input, charset)) {
		return std::string(charset);
	}
	return {};
}

} // namespace charencoding
} // namespace Newsboat

File: rust/libnewsboat-ffi/src/charencoding.rs
use libNewsboat::charencoding;

#[cxx::bridge(namespace = "Newsboat::charencoding::bridged")]
mod bridged {
    extern "Rust" {
        fn charset_from_bom(content: &[u8], output: &mut String) -> bool;
        fn charset_from_xml_declaration(content: &[u8], output: &mut String) -> bool;
        fn charset_from_content_type_header(content: &[u8], output: &mut String) -> bool;
    }
}

// Temporarily ignore clippy lint until PR is merged:
// https://github.com/rust-lang/rust-clippy/pull/12756
#[allow(clippy::assigning_clones)]
fn charset_from_bom(content: &[u8], output: &mut String) -> bool {
    match charencoding::charset_from_bom(content) {
        Some(charset) => {
            *output = charset.to_owned();
            true
        }
        None => false,
    }
}

// Temporarily ignore clippy lint until PR is merged:
// https://github.com/rust-lang/rust-clippy/pull/12756
#[allow(clippy::assigning_clones)]
fn charset_from_xml_declaration(content: &[u8], output: &mut String) -> bool {
    match charencoding::charset_from_xml_declaration(content) {
        Some(charset) => {
            *output = charset.to_owned();
            true
        }
        None => false,
    }
}

// Temporarily ignore clippy lint until PR is merged:
// https://github.com/rust-lang/rust-clippy/pull/12756
#[allow(clippy::assigning_clones)]
fn charset_from_content_type_header(content: &[u8], output: &mut String) -> bool {
    match charencoding::charset_from_content_type_header(content) {
        Some(charset) => {
            *output = charset.to_owned();
            true
        }
        None => false,
    }
}
